Ensure installed/downloaded collections support the running ansible-core version

s-hertel opened the pull request:
##### SUMMARY

Fixes #78539

* Make ansible-galaxy consider existing `COLLECTIONS_ON_ANSIBLE_VERSION_MISMATCH`  configuration. Now collections with mismatched requirements are only installed/downloaded if it's set to "ignore".

##### ISSUE TYPE

- Bugfix Pull Request

s-hertel commented:
@jborean93 I'm in the middle of refactoring this to use `COLLECTIONS_ON_ANSIBLE_VERSION_MISMATCH` and a new general config to give a warning/error at runtime for undocumented requires_ansible, so please disregard until I get this out of draft state.

s-hertel commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -84,28 +88,44 @@ def _get_collection_versions(self, requirement: Requirement) -> t.Iterator[tuple
 
+    def _supports_ansible_version(self, requires_ansible : str | None) -> bool:
+        """Test if a requires_ansible value is compatible with the ansible-core version."""
+        if C.COLLECTIONS_ON_ANSIBLE_VERSION_MISMATCH == 'ignore':
+            return True
+        if requires_ansible is None:
+            return C.COLLECTIONS_ON_UNDOCUMENTED_ANSIBLE_VERSION == 'ignore'
+        return _does_collection_support_ansible_version(requires_ansible, ansible_version)
```
```suggestion
        if requires_ansible is None:
            return C.COLLECTIONS_ON_UNDOCUMENTED_ANSIBLE_VERSION == 'ignore'
        if C.COLLECTIONS_ON_ANSIBLE_VERSION_MISMATCH == 'ignore':
            return True
        return _does_collection_support_ansible_version(requires_ansible, ansible_version)
```

webknjaz commented:
/azp run

webknjaz commented on lib/ansible/config/base.yml:
```diff
@@ -222,13 +222,25 @@ COLLECTIONS_ON_ANSIBLE_VERSION_MISMATCH:
   name: Defines behavior when loading a collection that does not support the current Ansible version
   description:
   - When a collection is loaded that does not support the running Ansible version (with the collection metadata key `requires_ansible`).
+  - Starting in version '2.19', `ansible-galaxy` does not download or install collections with a mismatched version unless this option is set to 'ignore'.
```
Is this RST or MD?
```suggestion
  - Starting in version '2.19', ``ansible-galaxy`` does not download or install collections with a mismatched version unless this option is set to 'ignore'.
```

webknjaz commented on lib/ansible/galaxy/collection/concrete_artifact_manager.py:
```diff
@@ -605,6 +625,25 @@ def _normalize_galaxy_yml_manifest(
 
 
+@cache
+def _get_runtime_from_dir(b_path: bytes) -> t.Any:
```
Would `object` be enough?
```suggestion
def _get_runtime_from_dir(b_path: bytes) -> object:
```

s-hertel replied:
Fixed.

webknjaz commented on lib/ansible/galaxy/collection/concrete_artifact_manager.py:
```diff
@@ -338,6 +340,24 @@ def get_direct_collection_meta(self, collection):
+        elif collection.is_dir:
+            runtime = _get_runtime_from_dir(b_artifact_path) or {}
+        elif collection.is_virtual:
+            runtime = {}
```
Would this be better?
```suggestion
            runtime = {'requires_ansible': '*'}
```

s-hertel replied:
I don't think so. I'd still need to assume the dictionary could be empty below. The file is optional, and can be empty. I also want to avoid conflating the collection version (which can be "\*") with a collection's requires_ansible version (which cannot be "\*", but can be None an empty string or absent to match any ansible version).

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -84,28 +88,44 @@ def _get_collection_versions(self, requirement: Requirement) -> t.Iterator[tuple
+            return C.COLLECTIONS_ON_UNDOCUMENTED_ANSIBLE_VERSION == 'ignore'
+        if C.COLLECTIONS_ON_ANSIBLE_VERSION_MISMATCH == 'ignore':
+            return True
+        return _does_collection_support_ansible_version(requires_ansible, ansible_version)
```
Since this doesn't reference `self`, I'd move it out of the class.

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -84,28 +88,44 @@ def _get_collection_versions(self, requirement: Requirement) -> t.Iterator[tuple
         if not found_api and last_error is not None:
             raise last_error
 
+    def _supports_ansible_version(self, requires_ansible : str | None) -> bool:
```
nit: whitespace before the colon

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -84,28 +88,44 @@ def _get_collection_versions(self, requirement: Requirement) -> t.Iterator[tuple
         if not found_api and last_error is not None:
             raise last_error
 
+    def _supports_ansible_version(self, requires_ansible : str | None) -> bool:
```
I think, this would better belong in the `Requirement` / `Candidate` data classes. The API would be nicer with `requirement.is_compatible_with_current_runtime()` or something along those lines. They'd need to contain this bit of metadata.

Though, I think that pip has another neat trick in its resolver — as special data class that doesn't describe packages but runtime compatibility. So a `package_x` having `Python-Requires: >= 3.9` would translate into `PackageRequirement('package_x')` depending on `RuntimeRequirement('>= 3.9')`. And then, the resolver would just disregard dependency tree branches based on that…

Should we consider something along those lines?

s-hertel replied:
This metadata comes from the temporary artifact cache or Galaxy server, so the concrete artifact manager would need to be passed to the dataclass method. I don't love the existing pattern we have for `self._make_req_from_dict`, but I've gone ahead and added another.

Thanks to the pointer to Python-Requires. I think that's a good improvement since it allows the error message to be a result of the failure itself.

webknjaz commented on lib/ansible/galaxy/collection/galaxy_api_proxy.py:
```diff
@@ -84,28 +88,44 @@ def _get_collection_versions(self, requirement: Requirement) -> t.Iterator[tuple
+            for api, version in self._get_collection_versions(requirement):
+                versions += 1
+                requires_ansible = (api.requires_ansible.get(requirement.fqcn) or {}).get(version)
+                if self._supports_ansible_version(requires_ansible):
```
Could you explain why can't this be in `find_matches()` in providers?

webknjaz replied:
This doesn't feel right.. The API proxy abstraction was meant to be a thin layer for accessing different collection sources. But the more I read this, the more I'm convinced that this belongs more to the resolver mechanics.

s-hertel replied:
It's not that it can't be (now it is), but I had limited the scope of impact. The only caller of the API proxy abstraction get_collection_versions is in fact, find_matches. If there's only 1 candidate, filtering the collection versions based on that, or adding the requirement is a dependency, are functionally equivalent. One has much broader impact.

webknjaz replied:
That might be, but my concern was that if things start leaking across layers of abstraction, this would break the structure I've been envisioning all those years ago.

Anyway, since you've moved away from this, I'm marking this thread as resolved.

s-hertel replied:
I'm glad you consider your concern resolved despite our difference of opinion on that. This kind of thing is why it's more useful to bikeshed implementation details after reviewing functionality since it can make that kind of thing moot.

s-hertel committed:
ansible-galaxy - consider the requires_ansible metadata when determining which version of a collection is the best match

s-hertel committed:
Update type hint and add extra newline

s-hertel committed:
Update unit tests

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -1865,16 +1869,21 @@ def _resolve_depenency_map(
             max_rounds=2000000,  # NOTE: same constant pip uses
         ).mapping
     except CollectionDependencyResolutionImpossible as dep_exc:
```
Note that in https://github.com/ansible/ansible/pull/81607/files#diff-b1b46df93a963b73e0c4a4c7a598a35d3135076a5ecf23f1c07323be0005a5fc, I'm trying to make all the lib exceptions handled explicitly.

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -1865,16 +1869,21 @@ def _resolve_depenency_map(
+        conflict_causes = []
+        for req_inf in dep_exc.causes:
+            if req_inf.requirement.type == "requires_ansible":
+                collection = f"{req_inf.parent.fqcn!s}:{req_inf.parent.ver!s}"
```
This is equivalent:
```suggestion
                collection = str(req_inf.parent)
```

webknjaz commented on lib/ansible/galaxy/collection/concrete_artifact_manager.py:
```diff
@@ -762,3 +785,22 @@ def _tarfile_extract(
+
+
+@cache
+def _get_runtime_from_dir(b_path: bytes) -> object:
+    """Load the meta/runtime.yml from a collection directory."""
+    runtime_path = os.path.join(b_path, b"meta", b"runtime.yml")
+    with suppress(OSError):
+        with open(runtime_path, 'r') as fd:
+            return yaml_load(fd)
```
Can we plz start using pathlib for new things internally?
```suggestion
def _get_runtime_from_dir(b_path: bytes) -> object:
    """Load the meta/runtime.yml from a collection directory."""
    runtime_path = Path(to_native(b_path)) / "meta" / "runtime.yml"
    try:
        runtime_yml_txt = runtime_path.read_text(encoding="utf-8")
    except OSError:
        return None

    return yaml_load(runtime_yml_txt)
```

s-hertel replied:
Added.

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -184,6 +188,7 @@ def __init__(self, *args, **kwargs):
                 self.name,
                 self.ver
             )
+        self._parent = None
```
Could you explain the idea behind this?

s-hertel replied:
See https://github.com/ansible/ansible/pull/84715#discussion_r2114538788.  It wasn't working quite right last time you reviewed (incomplete) and it was easiest to fix by pilfering the dependency accounting out using a dedicated var instead of hanging it on requirements/candidates, so this specifically is no longer here.

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -641,3 +650,51 @@ def with_signatures_repopulated(self):  # type: (Candidate) -> Candidate
+        #   the collection documents it.
+        if self.ver in ('', None) and C.COLLECTIONS_ON_ANSIBLE_VERSION_MISMATCH != 'ignore':
+            warning = (
+                "" if self._parent.type != 'galaxy' or self._parent.src.api_server == "https://galaxy.ansible.com/api/"
```
Do we have a constant with this URL value anywhere?

s-hertel replied:
Now there is.

webknjaz commented on test/units/galaxy/test_collection_install.py:
```diff
@@ -82,6 +82,11 @@ def collection_artifact(request, tmp_path_factory):
         galaxy_obj.write(to_bytes(yaml.safe_dump(existing_yaml)))
         galaxy_obj.truncate()
 
+    os.mkdir(os.path.join(collection_path, 'meta'))
```
Note for the future: `tmp_path_factory` gives us pathlib objects. Rather than converting them to strings and bytes, we should use them directly instead of the low-level APIs.

webknjaz commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -450,13 +464,16 @@ def get_dependencies(self, candidate):
+            if not (self._with_deps or candidate.is_virtual):
+                continue
+            dependency = self._make_req_from_dict({'name': dep_name, 'version': dep_req})
+            dependency._parent = candidate  # Used by requires_ansible error handling, to display whether the error was caused by direct request or dependency
```
I think I'll need to play with this PR locally to understand this better, but the initial idea from pip that I mentioned earlier was that the depresolver would track such relationships already so we wouldn't need to do double-entry accounting.

s-hertel replied:
[ResolutionImpossible](https://github.com/sarugaku/resolvelib/blob/main/src/resolvelib/resolvers/exceptions.py#L47-L51) has a list [RequirementInformation](https://github.com/sarugaku/resolvelib/blob/main/src/resolvelib/structs.py#L27-L29) causes, which only gives access to the parent. I don't see a way to access the parent of the parent from the ResolutionImpossible object.

Because of this, injecting a fake requirement for collection metadata results in having no way to determine the origin of the collection containing the incompatible requires_ansible metadata from the ResolutionImpossible exception. My goal is to maintain the existing error message and append "requires Ansible ..." when it's the cause of failure.

On devel, a ResolutionImpossible error looks like:

```console
[ERROR]: Failed to resolve the requested dependencies map. Could not satisfy the following requirements:
* ns.dep:2.0.0 (direct request)
* ns.dep:1.0.0 (dependency of ns.col:1.0.0)
```

The error for the new requirement type could look like this without the accounting:

```console
[ERROR]: Failed to resolve the requested dependencies map. Could not satisfy the following requirements:
* Ansible:>=2.20 (dependency of ns.dep:1.0.0)
```

But I think this is better:

```console
[ERROR]: Failed to resolve the requested dependencies map. Could not satisfy the following requirements:
* ns.dep:1.0.0 (dependency of ns.col:1.0.0) requires Ansible >=2.20
```

Pip doesn't display this information via stderr, it's printed to stdout before the resolver fails. The pip error handling only uses the parent of the failed requirement for Python requirements, whereas our exception uses it already for all ResolutionImpossible requirements. Even if we add an opt-in debug mode to display it via stdout, imo it should be displayed by default in the case of failure.

webknjaz replied:
I see, I'll need to think about it and finally check it locally. I think I now understand the limitation you're trying to overcome, though.

And I agree that the last snippet looks best.

I'm a bit unclear as to why can't we just store references to virtual dependency parents in their requirement objects. For future access. That global mapping might not be necessary in such a case. Anyway, I'll try to get a feel of how this works locally and perhaps offer some suggestions once I understand more.

s-hertel replied:
> I'm a bit unclear as to why can't we just store references to virtual dependency parents in their requirement objects.

That was what I did initially. I put it back.

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -1865,16 +1869,21 @@ def _resolve_depenency_map(
+        for req_inf in dep_exc.causes:
+            if req_inf.requirement.type == "requires_ansible":
+                collection = f"{req_inf.parent.fqcn!s}:{req_inf.parent.ver!s}"
+                dep_origin = 'direct request' if req_inf.parent._parent is None else f'dependency of {req_inf.parent._parent!s}'
```
Can we just check if `parent` is of type `requires_ansible`? I'd expect that to be enough. Does anything prevent this from working? If yes, maybe there's something to be done about this in the resolver hooks.

s-hertel replied:
The `parent` is never type `requires_ansible`. It's a non-virtual collection requirement, which has a `requires_ansible` requirement as a dependency. The goal of this line is to determine the origin of the `parent`, which isn't accessible via the exception, as I've described in https://github.com/ansible/ansible/pull/84715#discussion_r2114538788.

Can you be clearer about what you think the issue is instead of just recommending something else? I'm confused at the suggestion and why you think there's an issue with the resolver hooks.

webknjaz replied:
I think I was trying to build a mental model of the changes you're proposing here and didn't understand the context fully. I probably confused the parent-child relationships originally and felt like there's something unnecessarily complicated happening here. But I've lost that context since. The more I'm reading this, the more I realize that just staring at the diff is not enough for me to give useful feedback, and I really do need to inspect this locally.

s-hertel replied:
@webknjaz Have you been able to take a closer look? I just rebased this branch since there was a merge conflict. Only the last couple commits are new since you last reviewed.

webknjaz replied:
I've been experimenting locally but never committed. Let me see if I can put that on top of you rebase..

s-hertel replied:
@webknjaz Since rebasing my PR causes you inconvenience, I'll closed this  as-is for posterity. Please review #86183 instead.

s-hertel committed:
Fix pilfering the dependency origin information

Exclude requires_ansible requirements that support ansible from the
error

Simplify a bit

s-hertel committed:
fix dependency comparison, clarify code comment

s-hertel committed:
Oops. Requirement objects are not hashable because signature_sources is a list, not a frozenset like Candidate.signatures. Use the repr to dedup instead relying on just the name and version for uniqueness

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -308,6 +309,9 @@ def __init__(
             self._cache = _load_cache(self._b_cache_path)
 
+        # For efficiency, this metadata is populated when listing collection versions.
+        self.requires_ansible = {}
```
Have you considered using a `defaultdict` here?
I think that it'll simplify entry creation on access below.
```suggestion
        self.requires_ansible = collections.defaultdict(dict)
```

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -893,7 +897,9 @@ def get_collection_versions(self, namespace, name):
-            versions += [v['version'] for v in data[results_key]]
+            for v in data[results_key]:
+                versions.append(v["version"])
+                self.requires_ansible.setdefault(f"{namespace}.{name}", {})[v["version"]] = v.get("requires_ansible")
```
With `defaultdict` suggested above, this line can look simpler:
```suggestion
                self.requires_ansible[f"{namespace}.{name}"][v["version"]] = v.get("requires_ansible")
```

webknjaz commented on lib/ansible/galaxy/dependency_resolution/providers.py:
```diff
@@ -425,13 +444,15 @@ def is_satisfied_by(self, requirement, candidate):
 
-    def get_dependencies(self, candidate):
-        # type: (Candidate) -> list[Candidate]
+    def get_dependencies(self, candidate: Candidate) -> t.Generator[Requirement]:
```
The `typing.Generator[YieldType, SendType, ReturnType]` generic has 3 arguments, I don't think that having a single value is valid.
However, it's usually unnecessary and using an iterator is a nicer alternative that I've been using everywhere:
```suggestion
    def get_dependencies(self, candidate: Candidate) -> t.Iterator[Requirement]:
```

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -527,7 +531,7 @@ def name(self):
     @property
     def canonical_package_id(self):
-        if not self.is_virtual:
+        if not self.is_virtual or self.is_ansible:
```
`if self.is_ansible` should probably generate a different string too (not '<virtual namespace ...>' but maybe '<ansible-core runtime>'. Otherwise, resolvelib would treat virtual namespaces as equal to runtimes.

```suggestion
        if self.is_ansible:
            return '<ansible-core runtime>'

        if not self.is_virtual:
```

s-hertel replied:
This expression is handling non-virtual candidates and runtimes the same way. Resolvelib won't treat non-virtual as equal to runtimes because 1) find_matches handles this type specifically 2) it cannot match an actual fqcn since there is no '.' in it.

In the last commit, I changed the canonical_package_id and displayed name from "Ansible" to "ansible-core".

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -537,7 +541,11 @@ def canonical_package_id(self):
+        return self.is_scm or self.is_subdirs or self.is_ansible
+
+    @property
+    def is_ansible(self):
```
I feel like this name needs a bit more accuracy.
```suggestion
    def is_ansible_core_runtime(self):
```

s-hertel replied:
I changed it to `is_ansible_runtime` in the last commit.

webknjaz commented on lib/ansible/galaxy/dependency_resolution/dataclasses.py:
```diff
@@ -537,7 +541,11 @@ def canonical_package_id(self):
     @property
     def is_virtual(self):
-        return self.is_scm or self.is_subdirs
+        return self.is_scm or self.is_subdirs or self.is_ansible
```
What does this influence? When I was designing the concept of virtual deps, I was imagining them corresponding to actual “packages” (collections). Adding the platform into the mix seems to taint that assumption, and I'm not sure if it'd cause problems in other places. That rely on this, corresponding to this ephemeral package concept.

s-hertel replied:
The virtual deps are collections, not the virtual namespace itself, and similarly, ansible-core is not a collection. Install/download automatically skip these, and it prevents reaching https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/dependency_resolution/providers.py#L422-L425 which would blow up for an ansible-core version that can't be interpreted as SemVer. I added a code comment:

```python
        # TODO: make it easier to plumb non-collection types through the dependency resolver.
        # For example, if is_virtual is False, it shouldn't be assumed that the requirement
        # is a collection, has metadata, has a SemVer version, and is installable.
        # It would be safer to handle expected types explicitly, and error otherwise.
```

webknjaz reviewed (commented):
These are mostly notes for myself, as I'm yet to deepen the understanding of the change by checking it locally.

s-hertel committed:
Put back import that got lost in rebase

Shuffle some code around, add a couple code comments

Fix type hint

Update tests

s-hertel committed:
Stash data on the requirement again

webknjaz commented:
Ack. I was wondering if you'd do that.. I'll make sure to rebase my experiments on top of that locally.
